app/vmalert/web_types.go

```diff
@@ -157,7 +157,7 @@ type apiRule struct {
 	// MaxUpdates is the max number of recorded ruleStateEntry objects
 	MaxUpdates int `json:"max_updates_entries"`
 	// Updates contains the ordered list of recorded ruleStateEntry objects
-	Updates []rule.StateEntry `json:"-"`
+	Updates []rule.StateEntry `json:"updates"`
```

Comment:
Please see this commit https://github.com/VictoriaMetrics/VictoriaMetrics/commit/e83f14210dfee627eb67302659952cd0467400ab:

>  * vmalert: exclude `updates` field from json marhsaling  This field isn't correctly marshaled right now. And implementing the correct marshaling for it doesn't seem right, since json representation is mostly used by systems like Grafana. And Grafana doesn't expect this field to be present.

Do you need this field? Can we conditionally marshal this field only if requested?

Reply:
Yes we need this field. I have hide this field from the json response if additional query param do not set.
Here is curl examples

```bash
curl  'http://localhost:8880/api/v1/rule' -d 'group_id=9057671247277533115' -d 'rule_id=12736842811613231495' -d 'with_field="1123"' | jq '.updates'
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   855  100   780  100    75  59374   5709 --:--:-- --:--:-- --:--:--  139k
null
```

```bash
curl  'http://localhost:8880/api/v1/rule' -d 'group_id=9057671247277533115' -d 'rule_id=12736842811613231495' -d '' | jq '.updates' 
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   838  100   780  100    58   104k   7980 --:--:-- --:--:-- --:--:--  409k
null
```

```bash
curl  'http://localhost:8880/api/v1/rule' -d 'group_id=9057671247277533115' -d 'rule_id=12736842811613231495' -d 'with_field="updates"' | jq '.updates'
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1412  100  1334  100    78  54466   3184 --:--:-- --:--:-- --:--:--  106k
[
  {
    "time": "2023-11-28T13:25:29.459307177Z",
    "at": "2023-11-28T13:24:00Z",
    "duration": 11350708,
    "error": null,
    "samples": 0,
    "series_fetched": 0,
    "curl": "curl -X POST -H 'Content-Type: application/json' 'http://vmdbselect:8481/select/0/prometheus/api/v1/query?query=changes%28process_start_time_seconds%7Bjob%3D~%22.%2A%28victoriametrics%7Cvmselect%7Cvminsert%7Cvmstorage%7Cvmagent%7Cvmalert%7Cvmsingle%7Cvmalertmanager%7Cvmauth%29.%2A%22%7D%5B15m%5D%29+%3E+2&step=300s&time=2023-11-28T13%3A24%3A00Z'"
  }
]
```

Reply:
> Do you need this field? I'd like to prevent sending all updates for rules to Grafana. Can we conditionally marshal this field only if requested?



> Please see this commit [e83f142](https://github.com/VictoriaMetrics/VictoriaMetrics/commit/e83f14210dfee627eb67302659952cd0467400ab):
> 
> > * vmalert: exclude `updates` field from json marhsaling  This field isn't correctly marshaled right now. And implementing the correct marshaling for it doesn't seem right, since json representation is mostly used by systems like Grafana. And Grafana doesn't expect this field to be present.
> 
> Do you need this field? I'd like to prevent sending all updates for rules to Grafana. Can we conditionally marshal this field only if requested?

 Btw, does grafana supports other extra fields like `DatasourceType` or it has issues only with `Updates` field? Probably, it's better to rename it for `StateUpdates` and introduce new field `Updates` later?

Reply:
> Btw, does grafana supports other extra fields like DatasourceType or it has issues only with Updates field?

It reads only fields it needs. Extra field will be ignored (at least at the moment when I write this comment).
So vmalert sends those extra fields you mentioned, except Updates field. Which I find much more expensive to send to grafana. Especially if someone will set the limit for Updates to big values.

Reply:
In this case, all fields not expected by grafana must be moved into separate `Extra` struct and joined only if caller supports it.

Reply:
I implemented this functionality when we have additional struct which can be added when it needed

Reply:
> In this case, all fields not expected by grafana must be moved into separate embedded Extra struct and joined only if caller supports it.

I wouldn't do it for all fields, as many of them were exposed for a long time already. This may break compatibility. 

But updates fields never was exposed. An comparing to other fields, it may contain arbitrary number of events, bloating the response significantly.